Gem install on Windows fails on Ruby 3.3.1 due to implicit function declaration

### Issue

On Windows 11, Ruby 3.3.1 using Windows Ruby Installer w/ Devkit, I encountered the following native build issue:

```
> gem install wdm
Temporarily enhancing PATH for MSYS/MINGW...
Building native extensions. This could take a while...
ERROR:  Error installing wdm:
        ERROR: Failed to build gem native extension.

    current directory: C:/ruby33/lib/ruby/gems/3.3.0/gems/wdm-0.1.1/ext/wdm
C:/ruby33/bin/ruby.exe extconf.rb
checking for -lkernel32... yes
checking for windows.h... yes
checking for ruby.h... yes
checking for HAVE_RUBY_ENCODING_H... yes
checking for rb_thread_call_without_gvl()... yes
creating Makefile

current directory: C:/ruby33/lib/ruby/gems/3.3.0/gems/wdm-0.1.1/ext/wdm
make DESTDIR\= sitearchdir\=./.gem.20240516-27944-207rfu sitelibdir\=./.gem.20240516-27944-207rfu clean

current directory: C:/ruby33/lib/ruby/gems/3.3.0/gems/wdm-0.1.1/ext/wdm
make DESTDIR\= sitearchdir\=./.gem.20240516-27944-207rfu sitelibdir\=./.gem.20240516-27944-207rfu
generating wdm_ext-x64-mingw-ucrt.def
compiling entry.c
compiling memory.c
compiling monitor.c
compiling queue.c
compiling rb_change.c
rb_change.c: In function 'extract_absolute_path_from_notification':
rb_change.c:139:5: warning: 'RB_OBJ_TAINT' is deprecated: taintedness turned out to be a wrong idea. [-Wdeprecated-declarations]
  139 |     OBJ_TAINT(path);
      |     ^~~~~~~~~
In file included from C:/ruby33/include/ruby-3.3.0/ruby/internal/core/rstring.h:30,
                 from C:/ruby33/include/ruby-3.3.0/ruby/internal/arithmetic/char.h:29,
                 from C:/ruby33/include/ruby-3.3.0/ruby/internal/arithmetic.h:24,
                 from C:/ruby33/include/ruby-3.3.0/ruby/ruby.h:28,
                 from C:/ruby33/include/ruby-3.3.0/ruby.h:38,
                 from wdm.h:22,
                 from rb_change.c:4:
C:/ruby33/include/ruby-3.3.0/ruby/internal/fl_type.h:824:1: note: declared here
  824 | RB_OBJ_TAINT(VALUE obj)
      | ^~~~~~~~~~~~
compiling rb_monitor.c
rb_monitor.c: In function 'rb_monitor_run_bang':
rb_monitor.c:509:29: error: implicit declaration of function 'rb_thread_call_without_gvl' [-Wimplicit-function-declaration]
  509 |         waiting_succeeded = rb_thread_call_without_gvl(wait_for_changes, monitor->process_event, stop_monitoring, monitor);
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
make: *** [Makefile:248: rb_monitor.o] Error 1

make failed, exit code 2

Gem files will remain installed in C:/ruby33/lib/ruby/gems/3.3.0/gems/wdm-0.1.1 for inspection.
Results logged to C:/ruby33/lib/ruby/gems/3.3.0/extensions/x64-mingw-ucrt/3.3.0/wdm-0.1.1/gem_make.out
```

### Workaround

Referring to this similar issue: https://github.com/ffi/ffi/issues/840

I found the following command works:

```
gem install wdm -- --with-cflags=-Wno-implicit-function-declaration
```

Yep,

did it as admin. got it successfully installed but the same error in vscode

`Error coming from `bundle install`

Command failed with exit code 5: bundle install
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

current directory:
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1/ext/wdm
C:/Ruby32-x64/bin/ruby.exe extconf.rb
checking for -lkernel32... yes
checking for windows.h... yes
checking for ruby.h... yes
checking for HAVE_RUBY_ENCODING_H... yes
checking for rb_thread_call_without_gvl()... yes
creating Makefile

current directory:
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1/ext/wdm
make DESTDIR\= sitearchdir\=./.gem.20240602-24924-6pt381
sitelibdir\=./.gem.20240602-24924-6pt381 clean

current directory:
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1/ext/wdm
make DESTDIR\= sitearchdir\=./.gem.20240602-24924-6pt381
sitelibdir\=./.gem.20240602-24924-6pt381
generating wdm_ext-x64-mingw-ucrt.def
compiling entry.c
compiling memory.c
compiling monitor.c
compiling queue.c
compiling rb_change.c
rb_change.c: In function 'extract_absolute_path_from_notification':
rb_change.c:139:5: warning: 'RB_OBJ_TAINT' is deprecated: taintedness turned out
to be a wrong idea. [-Wdeprecated-declarations]
  139 |     OBJ_TAINT(path);
      |     ^~~~~~~~~
In file included from
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/core/rstring.h:30,
from
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/arithmetic/char.h:29,
from
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/arithmetic.h:24,
                 from C:/Ruby32-x64/include/ruby-3.2.0/ruby/ruby.h:28,
                 from C:/Ruby32-x64/include/ruby-3.2.0/ruby.h:38,
                 from wdm.h:22,
                 from rb_change.c:4:
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/fl_type.h:116:30: note: declared
here
  116 | #define RB_OBJ_TAINT         RB_OBJ_TAINT
      |                              ^~~~~~~~~~~~
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/fl_type.h:852:1: note: in
expansion of macro 'RB_OBJ_TAINT'
  852 | RB_OBJ_TAINT(VALUE obj)
      | ^~~~~~~~~~~~
compiling rb_monitor.c
rb_monitor.c: In function 'rb_monitor_run_bang':
rb_monitor.c:509:29: error: implicit declaration of function
'rb_thread_call_without_gvl' [-Wimplicit-function-declaration]
509 |         waiting_succeeded = rb_thread_call_without_gvl(wait_for_changes,
monitor->process_event, stop_monitoring, monitor);
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
make: *** [Makefile:248: rb_monitor.o] Error 1

make failed, exit code 2

Gem files will remain installed in
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1
for inspection.
Results logged to
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/extensions/x64-mingw-ucrt/3.2.0/wdm-0.1.1/gem_make.out

  C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:125:in `run'
C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:51:in `block in
make'
  C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:43:in `each'
  C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:43:in `make'
C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/ext_conf_builder.rb:42:in
`build'
C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:193:in
`build_extension'
C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:227:in `block
in build_extensions'
  C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:224:in `each'
C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/ext/builder.rb:224:in
`build_extensions'
C:/Ruby32-x64/lib/ruby/site_ruby/3.2.0/rubygems/installer.rb:852:in
`build_extensions'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/rubygems_gem_installer.rb:72:in
`build_extensions'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/rubygems_gem_installer.rb:28:in
`install'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/source/rubygems.rb:207:in
`install'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/installer/gem_installer.rb:54:in
`install'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/installer/gem_installer.rb:16:in
`install_from_spec'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/installer/parallel_installer.rb:186:in
`do_install'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/installer/parallel_installer.rb:177:in
`block in worker_pool'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/worker.rb:62:in
`apply_func'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/worker.rb:57:in
`block in process_queue'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/worker.rb:54:in
`loop'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/worker.rb:54:in
`process_queue'
C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/bundler-2.3.20/lib/bundler/worker.rb:91:in
`block (2 levels) in create_threads'

An error occurred while installing wdm (0.1.1), and Bundler cannot continue.

In Gemfile:
  wdm
Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Using bundler 2.3.20
Using concurrent-ruby 1.1.10
Using ffi 1.15.4
Using rb-fsevent 0.11.1
Using bugsnag 6.24.2
Using rb-inotify 0.10.1
Using listen 3.7.1
Installing wdm 0.1.1 with native extensions
![Screenshot 2024-06-02 094154](https://github.com/Maher4Ever/wdm/assets/68458628/83ac78f4-6306-4d2d-813a-f953d7e9b7a8)
`

thanks for looking into it

I tried using this

gem install wdm -v '0.1.1' --install-dir C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0 -- --with-cflags="-O2 -pipe"


and got error again

gem install wdm -v '0.1.1' --install-dir C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0 -- --with-cflags="-O2 -pipe"
Temporarily enhancing PATH for MSYS/MINGW...
Building native extensions with: '--with-cflags=-O2 -pipe'
This could take a while...
ERROR:  Error installing wdm:
        ERROR: Failed to build gem native extension.

    current directory: C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1/ext/wdm
C:/Ruby32-x64/bin/ruby.exe extconf.rb --with-cflags\=-O2\ -pipe
checking for -lkernel32... yes
checking for windows.h... yes
checking for ruby.h... yes
checking for HAVE_RUBY_ENCODING_H... yes
checking for rb_thread_call_without_gvl()... yes
creating Makefile

current directory: C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1/ext/wdm
make DESTDIR\= sitearchdir\=./.gem.20240602-28864-mqcflf sitelibdir\=./.gem.20240602-28864-mqcflf clean

current directory: C:/Users/navdeep/AppData/Local/shopify-gems-nodejs/Cache/ruby/3.2.0/gems/wdm-0.1.1/ext/wdm
make DESTDIR\= sitearchdir\=./.gem.20240602-28864-mqcflf sitelibdir\=./.gem.20240602-28864-mqcflf
generating wdm_ext-x64-mingw-ucrt.def
compiling entry.c
compiling memory.c
compiling monitor.c
compiling queue.c
compiling rb_change.c
rb_change.c: In function 'extract_absolute_path_from_notification':
rb_change.c:139:5: warning: 'RB_OBJ_TAINT' is deprecated: taintedness turned out to be a wrong idea. [-Wdeprecated-declarations]
  139 |     OBJ_TAINT(path);
      |     ^~~~~~~~~
In file included from C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/core/rstring.h:30,
                 from C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/arithmetic/char.h:29,
                 from C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/arithmetic.h:24,
                 from C:/Ruby32-x64/include/ruby-3.2.0/ruby/ruby.h:28,
                 from C:/Ruby32-x64/include/ruby-3.2.0/ruby.h:38,
                 from wdm.h:22,
                 from rb_change.c:4:
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/fl_type.h:116:30: note: declared here
  116 | #define RB_OBJ_TAINT         RB_OBJ_TAINT
      |                              ^~~~~~~~~~~~
C:/Ruby32-x64/include/ruby-3.2.0/ruby/internal/fl_type.h:852:1: note: in expansion of macro 'RB_OBJ_TAINT'
  852 | RB_OBJ_TAINT(VALUE obj)
      | ^~~~~~~~~~~~
compiling rb_monitor.c
rb_monitor.c: In function 'rb_monitor_run_bang':
rb_monitor.c:509:29: error: implicit declaration of function 'rb_thread_call_without_gvl' [-Wimplicit-function-declaration]
  509 |         waiting_succeeded = rb_thread_call_without_gvl(wait_for_changes, monitor->process_event, stop_monitoring, monitor);
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
make: *** [Makefile:248: rb_monitor.o] Error 1

make failed, exit code 2

where should I look for non-vendored gem for wdm

--- a/Gemfile
+++ b/Gemfile
@@ -2,3 +2,10 @@ source 'https://rubygems.org'
 
 # Specify your gem's dependencies in wdm.gemspec
 gemspec
+
+group :development do
+  gem 'rake-compiler'
+  gem 'rspec'
+  gem 'guard-rspec'
+  gem 'guard-shell'
+end

--- a/Rakefile
+++ b/Rakefile
@@ -13,5 +13,5 @@ end
 
 desc "Open an irb session preloaded with WDM"
 task :console do
-  sh "irb -rubygems -I lib -r wdm"
+  sh "irb -I lib -r wdm"
 end

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,29 +1,32 @@
 ---
+image: Visual Studio 2022
 version: '{build}'
 
 skip_tags: true
 
 environment:
   matrix:
-    - ruby_version: "193"
-      devkit: C:\Ruby193\DevKit
-    - ruby_version: "200"
-      devkit: C:\Ruby21\DevKit
-    - ruby_version: "200-x64"
-      devkit: C:\Ruby21-x64\DevKit
-    - ruby_version: "21"
-      devkit: C:\Ruby21\DevKit
-    - ruby_version: "21-x64"
-      devkit: C:\Ruby21-x64\DevKit
-    - ruby_version: "22"
-      devkit: C:\Ruby21\DevKit
-    - ruby_version: "22-x64"
-      devkit: C:\Ruby21-x64\DevKit
+    - ruby_version: "head-x64"
+    - ruby_version: "25"
+    - ruby_version: "30"
+    - ruby_version: "33-x64"
+
+init:
+  - SET PATH=C:\Ruby%ruby_version%\bin;%PATH%
 
 install:
-  - SET PATH=C:\Ruby%ruby_version%\bin;%PATH%
-  - "%devkit%\\devkitvars.bat"
-  - gem install bundler --no-rdoc --no-ri
+  - ps: |
+      if ($env:ruby_version -like "*head*") {
+        $(new-object net.webclient).DownloadFile("https://github.com/oneclick/rubyinstaller2/releases/download/rubyinstaller-head/rubyinstaller-$env:ruby_version.exe", "$pwd/ruby-setup.exe")
+        cmd /c ruby-setup.exe /verysilent /currentuser /dir=C:/Ruby$env:ruby_version
+      }
+  - ruby --version
+  - gem --version
+  - ridk version
+  - ridk enable
+  - c:/msys64/usr/bin/bash -lc "pacman -S --noconfirm --needed ${MINGW_PACKAGE_PREFIX}-gcc"
+  - gcc -v
+  - gem install bundler:2.3.27 --conservative --no-doc
   - bundler env
   - bundle install --retry=3
 

--- a/ext/wdm/extconf.rb
+++ b/ext/wdm/extconf.rb
@@ -16,13 +16,10 @@ def windows?
 end
 
 if windows? and
-   have_library("kernel32") and
-   have_header("windows.h") and
-   have_header("ruby.h")    and
-   have_const('HAVE_RUBY_ENCODING_H')
+  have_library("kernel32") and
+  have_header("windows.h")
 then
-   have_func('rb_thread_call_without_gvl')
-   generate_makefile()
+  generate_makefile()
 else
-  generate_dummy_makefile() 
+  generate_dummy_makefile()
 end

--- a/ext/wdm/rb_change.c
+++ b/ext/wdm/rb_change.c
@@ -136,8 +136,6 @@ extract_absolute_path_from_notification(const LPWSTR base_dir, const PFILE_NOTIF
         multibyte_filepath_buffer_size - 1, // -1 because this func takes the chars count, not bytes count
         wdm_rb_enc_utf8);
 
-    OBJ_TAINT(path);
-
     return path;
 }
 

--- a/ext/wdm/rb_monitor.c
+++ b/ext/wdm/rb_monitor.c
@@ -99,6 +99,16 @@ monitor_free(LPVOID param)
     wdm_monitor_free(monitor);
 }
 
+static const rb_data_type_t monitor_data_type = {
+    .wrap_struct_name = "WDM::Monitor",
+    .function = {
+        .dmark = monitor_mark,
+        .dfree = monitor_free,
+        .dsize = NULL,
+    },
+    .flags = 0
+};
+
 static VALUE
 rb_monitor_alloc(VALUE self)
 {
@@ -106,7 +116,7 @@ rb_monitor_alloc(VALUE self)
     WDM_DEBUG("Allocating a new monitor object!");
     WDM_DEBUG("--------------------------------");
 
-    return Data_Wrap_Struct(self, monitor_mark, monitor_free, wdm_monitor_new());
+    return TypedData_Wrap_Struct(self, &monitor_data_type, wdm_monitor_new());
 }
 
 static DWORD
@@ -156,7 +166,7 @@ combined_watch(BOOL recursively, int argc, VALUE *argv, VALUE self)
     // TODO: Maybe raise a more user-friendly error?
     rb_need_block();
 
-    Data_Get_Struct(self, WDM_Monitor, monitor);
+    TypedData_Get_Struct(self, WDM_Monitor, &monitor_data_type, monitor);
 
     EnterCriticalSection(&monitor->lock);
         running = monitor->running;
@@ -467,7 +477,7 @@ rb_monitor_run_bang(VALUE self)
 
     WDM_DEBUG("Running the monitor!");
 
-    Data_Get_Struct(self, WDM_Monitor, monitor);
+    TypedData_Get_Struct(self, WDM_Monitor, &monitor_data_type, monitor);
     already_running = FALSE;
 
     EnterCriticalSection(&monitor->lock);
@@ -529,7 +539,7 @@ rb_monitor_stop(VALUE self)
 {
     WDM_PMonitor monitor;
 
-    Data_Get_Struct(self, WDM_Monitor, monitor);
+    TypedData_Get_Struct(self, WDM_Monitor, &monitor_data_type, monitor);
 
     stop_monitoring(monitor);
 

--- a/ext/wdm/rb_monitor.c
+++ b/ext/wdm/rb_monitor.c
@@ -50,7 +50,7 @@ static void CALLBACK handle_entry_change(DWORD, DWORD, LPOVERLAPPED);
 static BOOL register_monitoring_entry(WDM_PEntry);
 static DWORD WINAPI start_monitoring(LPVOID);
 
-static VALUE wait_for_changes(LPVOID);
+static void *wait_for_changes(void *);
 static void process_changes(WDM_PQueue);
 static void stop_monitoring(LPVOID);
 static VALUE rb_monitor_run_bang(VALUE);
@@ -367,14 +367,14 @@ start_monitoring(LPVOID param)
     return 0;
 }
 
-static VALUE
-wait_for_changes(LPVOID param)
+static void *
+wait_for_changes(void *param)
 {
-    HANDLE process_event;
+    HANDLE process_event = (HANDLE)param;
+    VALUE rb_res;
 
-    process_event = (HANDLE)param;
-
-    return WaitForSingleObject(process_event, INFINITE) == WAIT_OBJECT_0 ? Qtrue : Qfalse;
+    rb_res = WaitForSingleObject(process_event, INFINITE) == WAIT_OBJECT_0 ? Qtrue : Qfalse;
+    return (void *)rb_res;
 }
 
 static void
@@ -503,13 +503,7 @@ rb_monitor_run_bang(VALUE self)
 
     while ( monitor->running ) {
 
-        // Ruby 2.2 removed the 'rb_thread_blocking_region' function. Hence, we now need
-        // to check if the replacement function is defined and use it if it's available.
-        #ifdef HAVE_RB_THREAD_CALL_WITHOUT_GVL
-        waiting_succeeded = rb_thread_call_without_gvl(wait_for_changes, monitor->process_event, stop_monitoring, monitor);
-        #else
-        waiting_succeeded = rb_thread_blocking_region(wait_for_changes, monitor->process_event, stop_monitoring, monitor);
-        #endif
+        waiting_succeeded = (VALUE)rb_thread_call_without_gvl(wait_for_changes, monitor->process_event, stop_monitoring, monitor);
 
         if ( waiting_succeeded == Qfalse ) {
             rb_raise(eWDM_Error, "Failed while waiting for a change in the watched directories!");

--- a/ext/wdm/wdm.h
+++ b/ext/wdm/wdm.h
@@ -3,7 +3,7 @@
 // Support Windows 2000 and later,
 // this is needed for 'GetLongPathNameW' (both of the following defines)
 #ifndef WINVER
-#define WINVER 0x0500 
+#define WINVER 0x0500
 #endif
 #ifndef _WIN32_WINNT
 #define _WIN32_WINNT 0x0500
@@ -14,13 +14,14 @@
 #define WIN32_LEAN_AND_MEAN
 #endif
 #ifndef VC_EXTRALEAN
-#define VC_EXTRALEAN        
+#define VC_EXTRALEAN
 #endif
 
 #include <Windows.h>
 
 #include <ruby.h>
 #include <ruby/encoding.h>
+#include <ruby/thread.h>
 
 #ifndef WDM_H
 #define WDM_H

--- a/spec/support/fixture_helper.rb
+++ b/spec/support/fixture_helper.rb
@@ -21,7 +21,7 @@ module WDM
 
     ensure
       FileUtils.cd pwd
-      FileUtils.rm_rf(path) if File.exists?(path)
+      FileUtils.rm_rf(path) if File.exist?(path)
     end
   end
 end

--- a/spec/wdm/monitor_spec.rb
+++ b/spec/wdm/monitor_spec.rb
@@ -170,14 +170,6 @@ describe WDM::Monitor do
       end
     end
 
-    it 'marks changed paths as tainted' do
-      result = run_with_fixture(subject) do
-        touch 'file.txt'
-      end
-
-      expect(result.change.path).to be_tainted
-    end
-
     it 'reports changes with absolute paths even when passed relative directory to watch' do
       fixture do |dir|
         relative_dir = Pathname.new(dir).relative_path_from(Pathname.new(Dir.pwd)).to_s

--- a/wdm.gemspec
+++ b/wdm.gemspec
@@ -16,7 +16,7 @@ Gem::Specification.new do |gem|
   gem.require_paths = ['lib']
   gem.version       = '0.1.1'
 
-  gem.required_ruby_version = '>= 1.9.2'
+  gem.required_ruby_version = '>= 2.5'
 
   gem.add_development_dependency 'rake-compiler'
   gem.add_development_dependency 'rspec'

--- a/wdm.gemspec
+++ b/wdm.gemspec
@@ -17,11 +17,4 @@ Gem::Specification.new do |gem|
   gem.version       = '0.1.1'
 
   gem.required_ruby_version = '>= 2.5'
-
-  gem.add_development_dependency 'rake-compiler'
-  gem.add_development_dependency 'rspec'
-  gem.add_development_dependency 'guard-rspec'
-  gem.add_development_dependency 'guard-shell'
-  gem.add_development_dependency 'pry'
-  gem.add_development_dependency 'devkit'
 end

